### PR TITLE
perf: avoid unnecessary count on getManyAndCount

### PR DESCRIPTION
**Pull request by EQuincerot**
### Description of change

When `getManyAndCount` is called, stop execute the `count` query if it can be deduced from the
number of returned rows. This will **increase performance** by avoiding one round trip. It would be especially useful when using pagination libraries such as https://github.com/ppetzold/nestjs-paginate

#### Example

```js
const [entities, count] = await connection.manager
    .createQueryBuilder(Post, "post")
    .limit(10)
    .orderBy("post.id")
    .getManyAndCount()
```

Before this change, it triggers two queries:

```sql
SELECT * FROM post ORDER BY post.id LIMIT 10;
SELECT count(*) FROM post;
```
But if the first query returns only 5 results, we know that the count will also be 5.

After this change, it triggers only the first query:

```sql
SELECT * FROM post ORDER BY post.id LIMIT 10;
```

### Pull-Request Checklist

-   [x] Code is up-to-date with the `master` branch
-   [ ] This pull request links relevant issues as `Fixes #00000`
-   [x] There are new or updated unit tests validating the change
-   [ ] Documentation has been updated to reflect this change

### CODE REVIEW

File: test/other-issues/lazy-count/lazy-count.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                afterQuery.clear()
+
+                const [entities, count] = await connection.manager
+                    .createQueryBuilder(Post, "post")
+                    .limit(10)
+                    .orderBy("post.id")
+                    .getManyAndCount()
+
+                expect(count).to.be.equal(5)
+                expect(entities).not.to.be.undefined
```
Comment (alumni) [inline test/other-issues/lazy-count/lazy-count.ts]: `expect(entities).to.have.length(5)` should be enough to replace all checks on `entities`.

File: src/query-builder/SelectQueryBuilder.ts [hunk 2]
```diff
@@ -1875,11 +1875,28 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
...
+
+            let count
+            if (
+                ((this.expressionMap.take !== undefined &&
+                    entitiesAndRaw.entities.length < this.expressionMap.take) ||
+                    (this.expressionMap.limit !== undefined &&
+                        entitiesAndRaw.entities.length <
+                            this.expressionMap.limit)) &&
+                this.expressionMap.skip === undefined &&
+                this.expressionMap.offset === undefined
```
Comment (alumni) [inline src/query-builder/SelectQueryBuilder.ts]: There are some edge cases for `take`/`limit`/`skip`/`offset`: `0`, `null`, `undefined`.

File: test/other-issues/lazy-count/lazy-count.ts
Comment (alumni) [inline test/other-issues/lazy-count/lazy-count.ts:1]: Would be great to have a test for one-to-many / many-to-many relations (e.g. when a `Post` has many `Comment`s):

```ts
const postRepository = dataSource.getRepository(Post);

const [posts, postCount]  = await postRepository.findAndCount({
  relations: { comments: true },
  where: { content: Like('%test%') },
});
```

File: test/other-issues/lazy-count/lazy-count.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                schemaCreate: true,
+            })),
+    )
+    beforeEach(() => reloadTestingDatabases(connections))
+    after(() => closeTestingConnections(connections))
+
+    it("skip count query when less entities are returned than the limit", () =>
+        Promise.all(
+            connections.map(async function (connection) {
+                await savePostEntities(connection, 5)
```
Comment (alumni) [inline test/other-issues/lazy-count/lazy-count.ts:30]: `savePostEntities` does not exist?

Comment (EQuincerot) [reply test/other-issues/lazy-count/lazy-count.ts:30]: Not sure of what you mean here. It's a method that is defined at the end of the test. I created to mutualize the code creating entities, but if you prefer I can remove it.

Comment (alumni) [reply test/other-issues/lazy-count/lazy-count.ts:30]: Ah it's at the bottom, I didn't expect to see it there :)

**Review by alumni — APPROVED** [review]

**Review by gioboa — APPROVED** [review]

**Comment by gioboa** [comment]
Thanks for your amazing help @EQuincerot 
Kudos for your first PR in this project 🎉

File: src/query-builder/SelectQueryBuilder.ts [hunk 4]
```diff
@@ -1903,6 +1912,56 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
         }
     }
 
+    private lazyCount(entitiesAndRaw: { entities: Entity[]; raw: any[] }) {
```
Comment (sgarner) [inline src/query-builder/SelectQueryBuilder.ts]: Method should declare its return type:

```suggestion
    private lazyCount(entitiesAndRaw: { entities: Entity[]; raw: any[] }): number | undefined {
```

File: src/query-builder/SelectQueryBuilder.ts [hunk 5]
```diff
@@ -1875,11 +1875,20 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
...
-            // Creates a new cacheId for the count query, or it will retreive the above query results
-            // and count will return 0.
-            this.expressionMap.cacheId = cacheId ? `${cacheId}-count` : cacheId
-            const count = await this.executeCountQuery(queryRunner)
+
+            let count
+            let lazyCount = this.lazyCount(entitiesAndRaw)
+            if (lazyCount !== undefined) {
+                count = lazyCount
+            } else {
```
Comment (sgarner) [inline src/query-builder/SelectQueryBuilder.ts]: These variable declarations should be typed. And `lazyCount` should be a const since it's never updated.

In fact, we can simplify and lose the `lazyCount` variable entirely:

```suggestion
            let count: number | undefined = this.lazyCount(entitiesAndRaw)
            if (count === undefined) {
```

File: src/query-builder/SelectQueryBuilder.ts [hunk 6]
```diff
@@ -1903,6 +1912,56 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
...
+            this.expressionMap.take !== undefined &&
+            this.expressionMap.take !== null
+        if (hasTake) {
+            maxResults = this.expressionMap.take
+        }
+
+        if (hasLimit) {
+            // limit overrides take when no join is defined
+            maxResults = this.expressionMap.limit
+        }
```
Comment (sgarner) [inline src/query-builder/SelectQueryBuilder.ts]: Simpler with a ternary:

```suggestion
        const hasTake =
            this.expressionMap.take !== undefined &&
            this.expressionMap.take !== null
        // limit overrides take when no join is defined
        const maxResults = hasLimit
            ? this.expressionMap.limit
            : hasTake
            ? this.expressionMap.take
            : undefined
```

File: src/query-builder/SelectQueryBuilder.ts [hunk 7]
```diff
@@ -1903,6 +1912,56 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
...
+
+        if (
+            maxResults !== undefined &&
+            entitiesAndRaw.entities.length == maxResults
```
Comment (sgarner) [inline src/query-builder/SelectQueryBuilder.ts]:
```suggestion
            entitiesAndRaw.entities.length === maxResults
```

File: src/query-builder/SelectQueryBuilder.ts [hunk 8]
```diff
@@ -1903,6 +1912,56 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
...
+        }
+
+        if (
+            this.expressionMap.offset !== undefined &&
+            this.expressionMap.offset !== null &&
+            this.expressionMap.offset > 0
+        ) {
+            // offset overrides skip when no join is defined
+            previousResults = this.expressionMap.offset
+        }
```
Comment (sgarner) [inline src/query-builder/SelectQueryBuilder.ts]:
```suggestion
        const hasSkip = this.expressionMap.skip !== undefined &&
            this.expressionMap.skip !== null &&
            this.expressionMap.skip > 0
        const hasOffset = this.expressionMap.offset !== undefined &&
            this.expressionMap.offset !== null &&
            this.expressionMap.offset > 0
        // offset overrides skip when no join is defined
        const previousResults = hasOffset
            ? this.expressionMap.offset
            : hasSkip
            ? this.expressionMap.skip
            : 0;
```

File: src/query-builder/SelectQueryBuilder.ts [hunk 9]
```diff
@@ -1903,6 +1912,56 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
...
+        if (hasLimit) {
+            // limit overrides take when no join is defined
+            maxResults = this.expressionMap.limit
+        }
+
+        if (
+            maxResults !== undefined &&
+            entitiesAndRaw.entities.length == maxResults
+        ) {
+            return undefined
```
Comment (sgarner) [inline src/query-builder/SelectQueryBuilder.ts:1939]:
```suggestion
            // stop here when the result set contains the max number of rows; we need to execute a full count
            return undefined
```

File: test/other-issues/lazy-count/lazy-count.ts [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+                entities: [__dirname + "/entity/*{.js,.ts}"],
+                subscribers: [AfterQuerySubscriber],
+                dropSchema: true,
+                schemaCreate: true,
+            })),
+    )
+    beforeEach(() => reloadTestingDatabases(connections))
+    after(() => closeTestingConnections(connections))
+
+    it("skip count query when less entities are returned than the limit", () =>
```
Comment (sgarner) [inline test/other-issues/lazy-count/lazy-count.ts]:
```suggestion
    it("skip count query when fewer entities are returned than the limit", () =>
```

File: test/other-issues/lazy-count/lazy-count.ts [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+
+                expect(
+                    afterQuery
+                        .getCalledQueries()
+                        .filter((query) => query.match(/(count|cnt)/i)),
+                ).to.be.empty
+            }),
+        ))
+
+    it("skip count query when less entities are returned than the take", () =>
```
Comment (sgarner) [inline test/other-issues/lazy-count/lazy-count.ts]:
```suggestion
    it("skip count query when fewer entities are returned than the take", () =>
```

File: test/other-issues/lazy-count/lazy-count.ts [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+
+                expect(
+                    afterQuery
+                        .getCalledQueries()
+                        .filter((query) => query.match(/(count|cnt)/i)),
+                ).not.to.be.empty
+            }),
+        ))
+
+    it("skip count query when joining a subentity with a take", () =>
```
Comment (sgarner) [inline test/other-issues/lazy-count/lazy-count.ts]:
```suggestion
    it("skip count query when joining a relation with a take", () =>
```

File: test/other-issues/lazy-count/lazy-count.ts [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+
+                expect(
+                    afterQuery
+                        .getCalledQueries()
+                        .filter((query) => query.match(/(count|cnt)/i)),
+                ).to.be.empty
+            }),
+        ))
+
+    it("run count query when joining a subentity with a limit", () =>
```
Comment (sgarner) [inline test/other-issues/lazy-count/lazy-count.ts]:
```suggestion
    it("run count query when joining a relation with a limit", () =>
```

File: test/other-issues/lazy-count/lazy-count.ts [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+
+                expect(
+                    afterQuery
+                        .getCalledQueries()
+                        .filter((query) => query.match(/(count|cnt)/i)),
+                ).not.to.be.empty
+            }),
+        ))
+
+    it("skip count query when joining a subentity with a take and a skip", () =>
```
Comment (sgarner) [inline test/other-issues/lazy-count/lazy-count.ts]:
```suggestion
    it("skip count query when joining a relation with a take and a skip", () =>
```

File: test/other-issues/lazy-count/lazy-count.ts [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+
+                expect(
+                    afterQuery
+                        .getCalledQueries()
+                        .filter((query) => query.match(/(count|cnt)/i)),
+                ).to.be.empty
+            }),
+        ))
+
+    it("run count query when joining a subentity with a limit and an offset", () =>
```
Comment (sgarner) [inline test/other-issues/lazy-count/lazy-count.ts]:
```suggestion
    it("run count query when joining a relation with a limit and an offset", () =>
```

**Review by sgarner — CHANGES_REQUESTED** [review]
Nice fix, thank you @EQuincerot

Just a few code improvements that could be made.

File: src/query-builder/SelectQueryBuilder.ts (hunk 8 above)
Comment (EQuincerot) [reply src/query-builder/SelectQueryBuilder.ts]: I added type assertion to make typescript happy with the undefined that could actually not happen. I'm not 100% happy with that, if you have a better idea, feel free to share!

File: src/query-builder/SelectQueryBuilder.ts [hunk 16]
```diff
@@ -1903,6 +1909,55 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
...
+            this.expressionMap.offset !== undefined &&
+            this.expressionMap.offset !== null &&
+            this.expressionMap.offset > 0
+
+        // offset overrides skip when no join is defined
+        const previousResults: number = hasOffset
+            ? <number>this.expressionMap.offset
+            : hasSkip
+            ? <number>this.expressionMap.skip
+            : 0
```
Comment (sgarner) [inline src/query-builder/SelectQueryBuilder.ts]: Angle-bracket type assertions in TypeScript are generally discouraged these days.

I think you can achieve the effect better with a `!` non-null assertion:

```suggestion
        const previousResults: number = hasOffset
            ? this.expressionMap.offset!
            : hasSkip
            ? this.expressionMap.skip!
            : 0
```

**Review by sgarner — APPROVED** [review]
One more tweak, otherwise looks good :+1:

File: src/query-builder/SelectQueryBuilder.ts (hunk 16 above)
Comment (alumni) [reply src/query-builder/SelectQueryBuilder.ts]: These rules you are referring to are `consistent-type-assertions` and `non-nullable-type-assertion-style` from "stylistic" and "stylistic type checked" rulesets.

I was planning to progressively add some more eslint rulesets ("stylistic" and "strict"), just didn't want to enable all of them at once, thought about waiting a couple of releases between adding new rulesets, but it's good to follow them already if you are aware of them :)

**Review by sgarner — APPROVED** [review]
Thanks for your contribution @EQuincerot 💜

**Review by gioboa — APPROVED** [review]
Great improvement @EQuincerot thanks
